libs/entitlement/tools/securitymanager-scanner/src/main/java/org/elasticsearch/entitlement/tools/securitymanager/scanner/SecurityCheckClassVisitor.java

```diff
@@ -0,0 +1,262 @@
+        super(ASM9);
+        this.callerInfoByMethod = callerInfoByMethod;
+    }
+
+    @Override
+    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
+        super.visit(version, access, name, signature, superName, interfaces);
+        this.className = name;
+        this.classAccess = access;
+    }
+
+    @Override
+    public void visitSource(String source, String debug) {
+        super.visitSource(source, debug);
+        this.source = source;
+    }
+
+    @Override
+    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
+        if (excludedClasses.contains(this.className)) {
+            return super.visitMethod(access, name, descriptor, signature, exceptions);
+        }
+        return new SecurityCheckMethodVisitor(super.visitMethod(access, name, descriptor, signature, exceptions), name, access);
+    }
+
+    public void setCurrentModule(String moduleName, Set<String> moduleExports) {
+        this.moduleName = moduleName;
+        this.moduleExports = moduleExports;
+    }
+
+    public void setCurrentSourcePath(String path) {
+        this.sourcePath = path;
+    }
+
+    private class SecurityCheckMethodVisitor extends MethodVisitor {
+
+        private final String methodName;
+        private int line;
+        private String permissionType;
+        private String runtimePermissionType;
+        private final int methodAccess;
+
+        protected SecurityCheckMethodVisitor(MethodVisitor mv, String methodName, int methodAccess) {
+            super(ASM9, mv);
+            this.methodName = methodName;
+            this.methodAccess = methodAccess;
+        }
+
+        private static final Set<String> KNOWN_PERMISSIONS = Set.of("jdk.vm.ci.services.JVMCIPermission");
+
+        @SuppressForbidden(reason = "System.err is OK for this simple command-line tool")
+        private void handleException(String className, Throwable e) {
+            System.err.println("Cannot process " + className + ": " + e.getMessage());
+        }
+
+        @Override
+        public void visitTypeInsn(int opcode, String type) {
+            super.visitTypeInsn(opcode, type);
+            if (opcode == NEW) {
+                if (type.endsWith("Permission")) {
+                    var objectType = Type.getObjectType(type);
+                    if (KNOWN_PERMISSIONS.contains(objectType.getClassName())) {
+                        permissionType = type;
+                    } else {
+                        try {
+                            var clazz = Class.forName(objectType.getClassName());
+                            if (Permission.class.isAssignableFrom(clazz)) {
+                                permissionType = type;
+                            }
+                        } catch (ClassNotFoundException e) {
+                            handleException(objectType.getClassName(), e);
```

Comment:
Usually, for a helper that builds an exception to throw, you'd return that exception object and throw it from here. That way, the compiler knows that control exits the method at this point. So:

```
throw handleException(...);
```

Since this method doesn't do anything after this point, it doesn't matter here, but in other situations it can be the only way to avoid a compiler error about, say, variables being uninitialized.

Reply:
I'm not building an exception, I'm actually not throwing; I'm against swallowing exceptions in general, but this is part of the "extract additional information if possible, otherwise fine" code path.